File: 백준/Silver/13335. 트럭/트럭.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [n, w, L] = input[0].split(" ").map(Number);
//const [M] = input[0].split(" ").map(Number);
//const arr = input.slice(1).map(Number);

const map = input.slice(1).map((_) => _.trim().split(" ").map(Number));
map.pop();
//const [start, end] = input[M + 2].split(" ").map(Number);

function main() {
  const start = input[1].split(" ").map(Number);
  const middle = new Array(w).fill(0);
  const end = [];

  let answer = 0;

  while (end.length < n) {
    const curTruck = start[0];
    const outTruck = middle[0];

    //만약 다리의 맨 앞에 트럭이 존재하면 다리를 빠져나온다.
    if (outTruck !== 0) {
      end.push(outTruck);
    }

    //다리에 있는 트럭들 한 칸씩 앞으로 이동
    for (let i = 0; i < w - 1; i++) {
      middle[i] = middle[i + 1];
    }

    //다리의 맨 뒤는 비워진다.
    middle[w - 1] = 0;

    //다리의 길이, 다리의 최대 하중 조건을 모두 만족한다면
    if (
      middle.filter((el) => el !== 0).length < w &&
      middle.reduce((acc, cur) => acc + cur) + curTruck <= L
    ) {
      //출발하지 않은 트럭 하나를 다리의 맨 뒤로 보낸다.
      middle[w - 1] = curTruck;
      start.shift();
    }

    answer++;
  }

  console.log(answer);
}

main();

File: 백준/Silver/13335. 트럭/트럭.cc
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    //freopen("input.txt", "r", stdin);

    int n, w, L;
    cin >> n >> w >> L;

    int start[n];
    
    for(int i=0;i<n;i++) {
        cin >> start[i];
    }

    int middle[w];
    for(int i=0;i<w;i++) {
        middle[i] = 0;
    }

    int end = 0;
    int answer = 0;
    int curPos = 0;

    while(end < n) {
        int curTruck = start[curPos];
        int outTruck = middle[0];

        if(outTruck != 0) {
            end++;
        }

        for(int i=0;i<w-1;i++) {
            middle[i] = middle[i+1];
        }
        middle[w-1] = 0;

        int curLength = 0;
        int curWeight = 0;
        for(int i=0;i<w;i++) {
            if(middle[i] != 0) {
                curLength++;
            }
            curWeight += middle[i];
        }

        if(curLength < w && curWeight + curTruck <= L) {
            middle[w-1] = curTruck;
            curPos++;
        }

        answer++;
    }

    cout << answer << endl;

    return 0;
}
